sqldelight-runtime/src/main/java/com/squareup/sqldelight/SqlDelightDatabase.kt

```diff
@@ -1,93 +1,76 @@
  * A transaction-aware [SupportSQLiteOpenHelper] wrapper which can queue work to be performed after
  * the active [Transaction] ends with [afterTransaction].
  */
 abstract class SqlDelightDatabase(
-    internal val helper: SupportSQLiteOpenHelper
-) : Transactor, Closeable {
-
-  private val transactions = ThreadLocal<SqliteTransaction>()
-
-  private val transaction = object : Transaction {
-    override fun markSuccessful() {
-      helper.writableDatabase.setTransactionSuccessful()
-    }
-
-    override fun yieldIfContendedSafely(): Boolean {
-      return helper.writableDatabase.yieldIfContendedSafely()
-    }
-
-    override fun yieldIfContendedSafely(sleepAmount: Long, sleepUnit: TimeUnit): Boolean {
-      return helper.writableDatabase.yieldIfContendedSafely(sleepUnit.toMillis(sleepAmount))
-    }
-
-    override fun end() {
-      val transaction = transactions.get() ?: throw IllegalStateException("Not in transaction.")
-      transactions.set(transaction.parent)
-      helper.writableDatabase.endTransaction()
-      if (transaction.commit) {
-        transaction.forEach { function -> afterTransaction(function) }
-      }
-    }
-
-    override fun close() {
-      end()
-    }
-  }
+  internal val helper: SqlDatabaseHelper
+) {
+  private val transactions = ThreadLocal<Transaction>()
+  private val actions = ArrayList<() -> Unit>()
 
-  override fun newTransaction(): Transaction {
-    val transaction = SqliteTransaction(transactions.get())
-    transactions.set(transaction)
-    helper.writableDatabase.beginTransactionWithListener(transaction)
-
-    return this.transaction
-  }
-
-  override fun newNonExclusiveTransaction(): Transaction {
-    val transaction = SqliteTransaction(transactions.get())
-    transactions.set(transaction)
-    helper.writableDatabase.beginTransactionWithListenerNonExclusive(transaction)
-
-    return this.transaction
-  }
-
-  override fun afterTransaction(function: () -> Unit) {
+  fun afterTransaction(function: () -> Unit) {
     val transaction = transactions.get()
     if (transaction != null) {
-      transaction.add(function)
+      actions.add(function)
     } else {
       function()
     }
   }
 
-  override fun close() {
+  fun close() {
     helper.close()
   }
-}
 
-private class SqliteTransaction(
-    val parent: SqliteTransaction?
-) : LinkedHashSet<() -> Unit>(), SQLiteTransactionListener {
-  var commit: Boolean = false
+  fun transaction(body: Transaction.() -> Unit) {
+    val parent = transactions.get()
+    val transaction = Transaction()
 
-  override fun onBegin() {}
+    if (parent == null) {
+      helper.getConnection().beginTransaction()
+    }
 
-  override fun onCommit() {
-    commit = true
-  }
+    try {
+      transactions.set(transaction)
+      transaction.body()
+      transactions.set(parent)
+    } catch(e: RollbackException) {
+      if (parent != null) throw e
+      helper.getConnection().rollbackTransaction()
+      return
+    } catch (e: RuntimeException) {
```

Comment:
I'm not sure this is safe. It doesn't handle non-RuntimeException or Errors. This should probably be a `finally` block.